### PR TITLE
Fix softmax export

### PR DESCRIPTION
**Pull request by pengwa**
### Description

Why we need to define softmax export logic here?
For the usage `nn.functional.softmax(attn_weights, dim=-1, dtype=torch.float32)` in the model,
 https://github.com/huggingface/transformers/blob/76a33a10923ccc1074917f6b6a1e719e626b7dc9/src/transformers/models/mistral/modeling_mistral.py#L302
 If dtype is specified, the input tensor is casted to dtype before the operation is performed.
 This is useful for preventing data type overflows. While existing ONNX exporter do the cast after the operation, which is not correct. (https://github.com/pytorch/pytorch/blob/cf06189a2d2785ac493bcd0d55e520af5a0e3b97/torch/onnx/symbolic_opset13.py#L27).
 This override can be a workaround before PyTorch fix the issues in coming releases.
 (TODO: pengwa - add PyTorch versions when the issue is fixed).

@thiagocrepaldi We may need a fix in PyTorch repo as well.

### CODE REVIEW

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 1]
```diff
@@ -6496,3 +6497,56 @@ def run_step(model, x, y, z):
...
+        execution_mgr._debug_options.save_onnx_models.path,
+        _get_onnx_file_name(
+            execution_mgr._debug_options.save_onnx_models.name_prefix, "torch_exported", execution_mgr._export_mode
+        ),
+    )
+
+    onnx_model = onnx.load(path)
+    onnx_nodes = [n for n in onnx_model.graph.node]
+
+    assert onnx_nodes[0].op_type == "Cast"
```
Comment (zhijxu-MS) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:6548]: not quite sure, does onnx_model.graph.node guarantees the nodes will be in topo order?

Comment (pengwa) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:6548]: yes it should be in topo order, otherwise, when we onnx.load it, it will fail to found the producer's outputs. if not in the right order.

**Review by wschin — APPROVED** [review]

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 2]
```diff
@@ -6496,3 +6497,56 @@ def run_step(model, x, y, z):
...
+    )
+
+    def run_step(model, attn_weight):
+        prediction = model(attn_weight)
+        prediction.sum().backward()
+        return prediction
+
+    # reset manual seed to reset the generator
+    torch.manual_seed(2333)
+    attn_weight = torch.randn([20, 6, 10, 10], dtype=torch.float, device=device, requires_grad=True)
```
Comment (wschin) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:6524]: This test doesn't really test the overflow. Ideally, you can create tensor containing values close to fp16's max value, then check if output is not Inf/NaN.

Comment (pengwa) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:6524]: True, let me add this in later changes.
